_posts/2023-08-01-tmt-cross-project-testing.md

```diff
@@ -0,0 +1,131 @@
+
+## Packit and tmt to the rescue
+
+It turns out that modern [Packit](https://packit.dev/) and [tmt](https://tmt.readthedocs.io/) know enough tricks to actually do this. The high-level strategy is:
+
+ 1. Configure Packit to build every commit of your main branch, i.e. every merged PR, into a [COPR](https://copr.fedorainfracloud.org/) repository. That is not just useful for CI, but also for human users: They can easily test the latest fixes or features before you officially release them.
+
+ 2. Identify some strategic dependencies of your project which potentially or actually break your tests often.
+
+ 3. For each of these, add a tmt test plan for your project, and configure Packit to run it for each PR. These run as a separate status, so it makes it very clear if/when a proposed change in the dependency breaks your project.
+
+## RPM repository for each commit
+
+The prerequisite to run tests from your main branch is that they run against the corresponding code, and not against some random older version from some Linux distribution. For a project's own tests, packit builds rpms from a PR automatically, but you need to set this up yourself to test an external project.
+
+The canonical way to set this up is with a COPR. Go to <https://copr.fedorainfracloud.org> and create a "New Project" in your organization or user. Fill in the "1. Project information" section and mark your desired distro releases in "2. Build options". At the bottom, also fill in the "Packit allowed forge projects", i.e. allow the Packit service to interact with your project's pull requests. Example:
+
+![packit allowed forge projects](/images/packit-allowed-forge-projects.png)
+
+Take a look at [Cockpit's "main-builds" COPR](https://copr.fedorainfracloud.org/coprs/g/cockpit/main-builds/) for an example.
+
+Then configure `packit.yaml` in your project to do a [build for each commit](https://packit.dev/docs/configuration/upstream/copr_build#supported-triggers) that lands on your "main" branch; that may have a different name for your project of course. For example, [cockpit-podman did this](https://github.com/cockpit-project/cockpit-podman/pull/1365):
+
+```yaml
+- job: copr_build
+  trigger: commit
+  branch: "^main$"
+  owner: "@cockpit"
+  project: "main-builds"
+  preserve_project: True
+```
+
+After landing that, check your COPR's "Builds" page that you actually get a successful package build. The first run will fail, you will get a notification email and have to approve the "packit" COPR user to trigger builds in your COPR.
```

Comment:
maybe you could mention that to avoid this, it should be possible to edit the permissions via copr CLI directly before (and here is also a link in our docs for the whole "custom Copr repo" setup https://packit.dev/docs/configuration/upstream/copr_build#using-a-custom-copr-project):

`$ copr-cli edit-permissions --builder packit [--admin packit] <project>`

Reply:
Thanks! I added a sentence which links to that piece of packit docs, that's very useful for automation.